### PR TITLE
180924 fg

### PR DESCRIPTION
**Pull request by SunilSimha**
Functions for running the Photo-z DM halos analysis.

### CODE REVIEW

File: frb/analysis/dmhalos_photoz.py [hunk 1]
```diff
@@ -0,0 +1,615 @@
+import numpy as np, os, glob
```
Comment (profxj) [inline frb/analysis/dmhalos_photoz.py:1]: I feel like this module might go best in `halos`
Thoughts?

File: frb/analysis/dmhalos_photoz.py [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+    Download photometry for galaxies within an FRB field.
+    Args:
+        coords (SkyCoord): Coordinates of the center of a cone search.
+        radius (Quantity, optional): Radius of cone search.
+        starbright (float, optional): Lower limit of r band mag. Objects brighter
+                                    than this will be removed.
+        starflagval (float, optional): Upper limit for a morphology-based classifier
+                                       flag. Objects more point-like (i.e. higher value)
+                                       will be filtered out.
+        gaicat (str, optional): Optional file with gaia catalog of stars within the same search
```
Comment (profxj) [inline frb/analysis/dmhalos_photoz.py:47]: Specify the format and columns of this table.

File: frb/analysis/dmhalos_photoz.py [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+    
+    return photom_cat
+
+def _gen_eazy_tab(photom_cat:Table, input_dir:str="eazy_in", name:str="FRB180924", out_dir="eazy_out", output_tab="no_stars_eazy.fits")->Table:
+    """
+    Run EAZY on the photometry and produce p(z) estimates.
+    Args:
+        photom_cat (Table): Photometry catalog.
+        input_dir (str, optional): Folder where EAZY config files are written.
+        name (str, optional): frb name
```
Comment (profxj) [inline frb/analysis/dmhalos_photoz.py:97]: explain what this is used for.

File: frb/analysis/dmhalos_photoz.py [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+        out_dir (str, optional): Folder where EAZY output is stored.
+        output_tab (str, optional): Name of the output summary table fits file.
+    Returns:
+        joined_tab (Table): EAZY results table conjoined with photom_cat.
```
Comment (profxj) [inline frb/analysis/dmhalos_photoz.py:101]: joined with which method?   e.g., `inner`, `outer`, ..

File: frb/analysis/dmhalos_photoz.py [hunk 5]
```diff
@@ -0,0 +1,615 @@
...
+        n_z (int, optional): Number of redshift grid points.
+        cigale_input (str, optional): Nameof input file to be produced.
+    Returns:
+        stacked_photom (Table): A table with multiple groups, one for each galaxy.
+            Each entry in a group has the same photometry but different redshift values.
+            This way, CIGALE can be run on the same galaxy at multiple redshift guesses
+            in one go.
+    """
+    # Define z values
+    z_range = np.linspace(zmin, zmax, n_z) # Go up in steps of 0.01
```
Comment (profxj) [inline frb/analysis/dmhalos_photoz.py:143]: Doesn't look like steps of `0.01` to me.

File: frb/analysis/dmhalos_photoz.py [hunk 6]
```diff
@@ -0,0 +1,615 @@
...
+
+    photom_cat['redshift'] = z_range[0] # Set up initial redshift value
+    photom_cat['ID'] = photom_cat['ID'].astype(str) # Convert form int to str
+    photom_cat.sort("separation")
+    photom_cat['ID'] = [ID.zfill(5)+"_{:0.2f}".format(z_range[0]) for ID in photom_cat['ID']]
+
+    # Create new table
```
Comment (profxj) [inline frb/analysis/dmhalos_photoz.py:150]: Whoa, there is surely a faster way to do this, but ok.

File: frb/analysis/dmhalos_photoz.py [hunk 7]
```diff
@@ -0,0 +1,615 @@
...
+        func = lambda x: halomass_from_stellarmass(x, z = z, randomize=True)
+        log_mhalo_array = p.map(func, log_mstar)
+        
+        return log_mhalo_array
+
+
+    # Loop over log_mstar:
+    for idx, log_mstar in enumerate(log_mstar_array):
+        temp_log_mstar = np.full(n_halo, log_mstar)
+        log_mhalo_array[idx] = mhalo_factory(temp_log_mstar, z = z, n_cores = 10)
```
Comment (profxj) [inline frb/analysis/dmhalos_photoz.py:314]: Make `n_cores` a parameter.

File: frb/analysis/dmhalos_photoz.py [hunk 8]
```diff
@@ -0,0 +1,615 @@
...
+    
+    # Store this in a fits file
+    prihdu  = fits.PrimaryHDU()
```
Comment (profxj) [inline frb/analysis/dmhalos_photoz.py:317]: I tend to prefer `numpy` or `hdf5` files these
days unless the extras (e.g. Header) of FITS
is warranted.

File: frb/analysis/dmhalos_photoz.py [hunk 9]
```diff
@@ -0,0 +1,615 @@
...
+    Returns:
+        mhalo_reals (np.ndarray): log_mhalo realizations.
+    """
+
+    # First produce realizations of mstar from a normal distribution.
+    mstar_reals = np.random.normal(log_mstar, log_mstar_err, n_mstar)
+
+    # Then get mean values of halo masses for each stellar mass.
+    mean_mhalo_reals = mean_interp(mstar_reals, z)
+    mean_mhalo_reals[mean_mhalo_reals>12.8] = 12.8 # Set a cutoff for the mean halo mass
```
Comment (profxj) [inline frb/analysis/dmhalos_photoz.py:368]: Just so you know, there is a `numpy.maximum` method that does this.

File: frb/analysis/dmhalos_photoz.py [hunk 10]
```diff
@@ -0,0 +1,615 @@
...
+    Produce DM estimates for a 3D grid of
+    redshift, offsets and log_halo_masses and write
+    them to disk.
+    Args:
+        frb_z(float): frb redshift
+        outdir(str, optional): data directory to store results
+        outfile(str, optional): name of results fits file (within outdir).
+    """
+    # Redshift grid
+    n_z = 100
```
Comment (profxj) [inline frb/analysis/dmhalos_photoz.py:441]: I think these should parameters to the method.

File: frb/analysis/dmhalos_photoz.py [hunk 11]
```diff
@@ -0,0 +1,615 @@
...
+    n_m = 100
+    log_halo_masses = np.linspace(8, 16, n_m)
+
+    ZZ, OO, MM = np.meshgrid(redshifts, offsets, log_halo_masses, indexing='ij')
+    raveled_z = ZZ.ravel()
+    raveled_o = OO.ravel()
+    raveled_m = MM.ravel()
+
+    def halo_dm(idx):
+        if raveled_m[idx] > 12.8:
```
Comment (profxj) [inline frb/analysis/dmhalos_photoz.py:458]: I know we had good reasons for this in our project, but
how about others?  Should `12.8` at least be a parameter?

**Review by profxj — CHANGES_REQUESTED** [review]
Great stuff.  Just a few requests..

Thx!

File: frb/analysis/dmhalos_photoz.py (hunk 1 above)
Comment (SunilSimha) [reply frb/analysis/dmhalos_photoz.py:1]: Makes sense. I shall move it there.

File: frb/analysis/dmhalos_photoz.py (hunk 2 above)
Comment (SunilSimha) [reply frb/analysis/dmhalos_photoz.py:47]: Done.

File: frb/analysis/dmhalos_photoz.py (hunk 3 above)
Comment (SunilSimha) [reply frb/analysis/dmhalos_photoz.py:97]: Done.

File: frb/analysis/dmhalos_photoz.py (hunk 4 above)
Comment (SunilSimha) [reply frb/analysis/dmhalos_photoz.py:101]: Done.

File: frb/analysis/dmhalos_photoz.py (hunk 5 above)
Comment (SunilSimha) [reply frb/analysis/dmhalos_photoz.py:143]: Removed this comment.

File: frb/analysis/dmhalos_photoz.py (hunk 7 above)
Comment (SunilSimha) [reply frb/analysis/dmhalos_photoz.py:314]: Done.

File: frb/analysis/dmhalos_photoz.py (hunk 8 above)
Comment (SunilSimha) [reply frb/analysis/dmhalos_photoz.py:317]: Noted and changed `.fits` to `.npz` wherever possible.

File: frb/analysis/dmhalos_photoz.py (hunk 9 above)
Comment (SunilSimha) [reply frb/analysis/dmhalos_photoz.py:368]: Thanks! Replaced with the `np.minimum`!

File: frb/analysis/dmhalos_photoz.py (hunk 10 above)
Comment (SunilSimha) [reply frb/analysis/dmhalos_photoz.py:441]: Changed.

File: frb/analysis/dmhalos_photoz.py (hunk 11 above)
Comment (SunilSimha) [reply frb/analysis/dmhalos_photoz.py:458]: Fair point. Introduced a parameter: `max_log_mhalo` where necessary.

File: frb/halos/photoz.py [hunk 12]
```diff
@@ -1,34 +1,32 @@
...
-from pathos.multiprocessing import ProcessingPool as Pool
-import progressbar
-
-from matplotlib import pyplot as plt
-
-from threedhst import eazyPy as ez
-
-from importlib import reload
+try:
+    from pathos.multiprocessing import ProcessingPool as Pool
```
Comment (profxj) [inline frb/halos/photoz.py:19]: Can you add these to the optional items to install
in our docs?

**Review by profxj — APPROVED** [review]

File: frb/halos/photoz.py (hunk 12 above)
Comment (SunilSimha) [reply frb/halos/photoz.py:19]: Done.

**Pull request merged**